### PR TITLE
Adding web page digest function to service module

### PR DESCRIPTION
**Pull request by ZiTao-Li**
---
Adding web page digest function to service module

---
## Description

As there are some recent internal requests about parsing the webpage, this PR introduce a webpage digestion service method to the framework.

* If a LLM is provided as the `model` parameter,  the webpage will first be split (by langchain_text_splitters.HTMLHeaderTextSplitter) and analyzed by LLM one by one.

* If there is no LLM provide, then the langchain_community.document_transformers.BeautifulSoupTransformer will be used to clean the webpage.

## Checklist

Please check the following items before code is ready to be reviewed.

- [x]  Code has passed all tests
- [x]  Docstrings have been added/updated in Google Style
- [x]  Documentation has been updated
- [x]  Code is ready for review

### CODE REVIEW

File: src/agentscope/service/web_search/web_digest.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    Parses the web page
+    :param url: the url of the web page
+    :param model: the model that is used to digest the web page content.
+    :param html_split_levels:  required parameter for HTMLHeaderTextSplitter
+    :param digest_prompt: prompt for the model to analyze the webpage content
+
+    :return: a ServiceResponse containing the digested content
+
+    If the model is None, then the webpage content will be processed
+    by **BeautifulSoupTransformer**.
```
Comment (DavdGao) [inline src/agentscope/service/web_search/web_digest.py:42]: In the docstring, it's unclear how the model and beautifulsoup4 will handle the web pages, e.g. what's the output of this function?

**Review by DavdGao — COMMENTED** [review]
1. Since there's no overlap in the processes for handling web pages via a model and a third-party library, they should be split into two separate methods like `bing_search` and `google_search`. The developer should decide which to use when choosing the service function. 
2. Please see inline comments.

File: src/agentscope/service/web_search/web_digest.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+from agentscope.service.service_response import ServiceResponse
+from agentscope.service.service_status import ServiceExecStatus
+from agentscope.models.model import ModelWrapperBase
+from agentscope.message import Msg
+
+
+DEFAULT_SYS_PROMPT = (
+    "You're a data cleaner. You job is to extract useful "
+    "training data from html or webpage source.\n"
```
[the remark this answers is not in the document]
Comment (ZiTao-Li) [reply src/agentscope/service/web_search/web_digest.py]: removed

File: src/agentscope/service/web_search/web_digest.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+
+def webpage_digest(
+    url: str,
+    model: Optional[ModelWrapperBase] = None,
+    html_split_levels: Optional[list[tuple[str, str]]] = None,
+    digest_prompt: Optional[str] = DEFAULT_SYS_PROMPT,
+) -> ServiceResponse:
+    """
+    Parses the web page
```
[the remark this answers is not in the document]
Comment (ZiTao-Li) [reply src/agentscope/service/web_search/web_digest.py]: Updated.

File: src/agentscope/service/web_search/web_digest.py (hunk 1 above)
Comment (ZiTao-Li) [reply src/agentscope/service/web_search/web_digest.py]: updated

File: src/agentscope/service/web_search/web_digest.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    :param url: the url of the web page
+    :param model: the model that is used to digest the web page content.
+    :param html_split_levels:  required parameter for HTMLHeaderTextSplitter
```
[the remark this answers is not in the document]
Comment (ZiTao-Li) [reply src/agentscope/service/web_search/web_digest.py]: Hide this parameter from users. Users now just need to specify which levels of tags on which they want to split the html, e.g., ("h1", "h2")

File: src/agentscope/service/web_search/web_digest.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+    "content of a webpage.\n"
+)
+
+DEFAULT_SPLIT_LEVEL = [
+    ("h1", "h1"),
+    ("h2", "h1"),
+]
+
+
+def webpage_digest(
```
Comment (pan-x-c) [inline src/agentscope/service/web_search/web_digest.py:29]: Extract a `load_webpage` method from this method.
It focuses on getting web pages and cleaning them without calling LLM.
Agents can use `load_webpage` to obtain the cleaned web page.

File: src/agentscope/service/web_search/web_digest.py [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+                        else:
+                            text += child.text
+                return " ".join(text.split())
+
+            text_parts = ""
+            if html_selected_tags is None:
+                html_selected_tags = ["h", "p", "li", "div", "a"]
+            for element in doc.find_all(recursive=True):
+                if element.name in html_selected_tags:
+                    text_parts += get_navigable_strings(element).strip(" \n\t")
```
Comment (pan-x-c) [inline src/agentscope/service/web_search/web_digest.py:252]: If the tags here are nested, will it cause the same content to be extracted multiple times?

File: src/agentscope/service/web_search/web_digest.py [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+        r"(?::\d+)?"  # optional port
+        r"(?:/?|[/?]\S+)$",
+        re.IGNORECASE,
+    )  # rest of url
+    return re.match(regex, url) is not None
+
+
+def web_load(
+    url: str,
+    html_parsing_types: Sequence[HTML_PARSING_TYPES] = tuple(
```
Comment (pan-x-c) [inline src/agentscope/service/web_search/web_digest.py:55]: It would be better to make these tags into bool type switches

**Review by DavdGao — COMMENTED** [review]
please see inline comments

File: tests/web_digest_test.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+        """test web_load function loading html"""
+        # Set up the mock response
+        mock_response = Mock()
+        mock_return_text = """
+            <!DOCTYPE html>
+            <html>
+            <body>
+                <div>
+                    <p>
+                    Some intro text about Foo. <a href="xxx">Examples</a>
```
Comment (pan-x-c) [inline tests/web_digest_test.py:28]: the <a> tag is not included in the expected result

Comment (ZiTao-Li) [reply tests/web_digest_test.py:28]: discussed in group chat.

File: src/agentscope/service/web_search/web_digest.py (hunk 5 above)
Comment (ZiTao-Li) [reply src/agentscope/service/web_search/web_digest.py]: Done

File: src/agentscope/service/web_search/web_digest.py (hunk 7 above)
Comment (ZiTao-Li) [reply src/agentscope/service/web_search/web_digest.py]: Updated to controlled by three parameters:
* `keep_raw`: if true, ServiceResponse.content contains raw html str
* `html_selected_tags`: if is not None nor empty, ServiceResponse.content contains the text extracted from those tags.
* `html_parse_func`: if not None, ServiceResponse.content has the return processed by the `html_parse_func`. `html_parse_func` is expected to take the html str as input.

File: src/agentscope/service/web_search/web_digest.py [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+    """
+    Use regex to check if a URL is valid
+    Args:
+        url (str): string to be checked
+
+    Returns:
+        bool: True if url is valid, False otherwise
+    """
+    # This regex pattern is designed to match most common URLs.
+    regex = re.compile(
```
[the remark this answers is not in the document]
Comment (ZiTao-Li) [reply src/agentscope/service/web_search/web_digest.py]: updated

File: src/agentscope/service/web_search/web_digest.py [hunk 10]
```diff
@@ -0,0 +1,298 @@
...
+                    f"Unsupported content type ({content_type}) "
+                    f"with url: ({url})",
+                )
+        else:
+            logger.warning(
+                f"Fail to load web page, "
+                f"status code {response.status_code}",
+            )
+            return ServiceResponse(
+                ServiceExecStatus.SUCCESS,
```
[the remark this answers is not in the document]
Comment (ZiTao-Li) [reply src/agentscope/service/web_search/web_digest.py]: Corrected.

File: src/agentscope/service/web_search/web_digest.py [hunk 11]
```diff
@@ -0,0 +1,298 @@
...
+        r"localhost|"  # localhost...
+        r"\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})"  # ...or ip
+        r"(?::\d+)?"  # optional port
+        r"(?:/?|[/?]\S+)$",
+        re.IGNORECASE,
+    )  # rest of url
+    return re.match(regex, url) is not None
+
+
+def web_load(
```
[the remark this answers is not in the document]
Comment (ZiTao-Li) [reply src/agentscope/service/web_search/web_digest.py]: Updated to controlled by three parameters:
* `keep_raw`: if true, ServiceResponse.content contains raw html str
* `html_selected_tags`: if is not None nor empty, ServiceResponse.content contains the text extracted from those tags.
* `html_parse_func`: if not None, ServiceResponse.content has the return processed by the `html_parse_func`. `html_parse_func` is expected to take the html str as input.

File: src/agentscope/service/__init__.py [hunk 12]
```diff
@@ -54,4 +55,7 @@ def get_help() -> None:
     "summarization",
     "retrieve_from_list",
     "ServiceExecStatus",
+    "webpage_digest",
```
[the remark this answers is not in the document]
Comment (ZiTao-Li) [reply src/agentscope/service/__init__.py]: updated

File: src/agentscope/service/web_search/web_digest.py [hunk 13]
```diff
@@ -0,0 +1,298 @@
...
+                # pylint: disable=cell-var-from-loop
+                text = ""
+                for child in e.children:
+                    if isinstance(child, Tag):
+                        # pylint: disable=cell-var-from-loop
+                        text += get_navigable_strings(child).strip(" \n\t")
+                    elif isinstance(child, NavigableString):
+                        if (e.name == "a") and (href := e.get("href")):
+                            if is_valid_url(href):
+                                text += f"{child.strip()} ({href})"
```
[the remark this answers is not in the document]
Comment (ZiTao-Li) [reply src/agentscope/service/web_search/web_digest.py]: updated

File: src/agentscope/service/web_search/web_digest.py [hunk 14]
```diff
@@ -0,0 +1,272 @@
...
+    """
+    header = {
+        "Accept-Language": "zh-CN,zh;q=0.9,en;q=0.8,en-GB;q=0.7,en-US;q=0.6",
+        "Cache-Control": "max-age=0",
+        "Upgrade-Insecure-Requests": "1",
+        "User-Agent": "Mozilla/5.0 (X11; Linux x86_64)"
+        " AppleWebKit/537.36 (KHTML, like Gecko) ",
+    }
+    try:
+        response = requests.get(url=url, headers=header, timeout=5)
```
Comment (pan-x-c) [inline src/agentscope/service/web_search/web_digest.py:94]: make `timeout` user configurable

File: src/agentscope/service/web_search/web_digest.py [hunk 15]
```diff
@@ -0,0 +1,298 @@
...
+                ServiceExecStatus.SUCCESS,
+                content="",
+            )
+    except Exception as e:
+        logger.warning(e)
+        return ServiceResponse(ServiceExecStatus.ERROR, content="")
+
+
+def html_parse(
+    html_response: requests.Response,
```
[the remark this answers is not in the document]
Comment (ZiTao-Li) [reply src/agentscope/service/web_search/web_digest.py]: updated

File: src/agentscope/service/web_search/web_digest.py (hunk 14 above)
Comment (ZiTao-Li) [reply src/agentscope/service/web_search/web_digest.py]: updated

File: src/agentscope/service/web_search/web_digest.py [hunk 16]
```diff
@@ -0,0 +1,273 @@
...
+        with `content` field is a dict, where keys are subset of
+        {"raw", "self_define_func", "selected_tags_text"}
+         For example, `content` field is
+        {
+            "raw": raw_content_of_web,
+            "selected_tags_text":
+                processed_text_content_of_the_selected_tags,
+        }
+
+    Returns:
```
Comment (DavdGao) [inline src/agentscope/service/web_search/web_digest.py:75]: Two returns in docstring.

File: src/agentscope/service/web_search/web_digest.py [hunk 17]
```diff
@@ -0,0 +1,273 @@
...
+    html_text: str,
+    keep_raw: bool = True,
+    html_selected_tags: Optional[Sequence[str]] = None,
+    html_parse_func: Optional[Callable] = None,
+) -> ServiceResponse:
+    """
+    Parse the obtained HTML file.
+    Args:
+        html_text (str):
+            HTTP response text
```
Comment (DavdGao) [inline src/agentscope/service/web_search/web_digest.py:154]: "HTTP response text" maybe confused. Is it the HTML source code?

File: src/agentscope/service/web_search/web_digest.py [hunk 18]
```diff
@@ -0,0 +1,273 @@
...
+            stored with key "raw".
+        html_selected_tags (Optional[Sequence[str]]):
+            the text in elements of `selected_tags` will
+            be extracted and stored with "selected_tags_text"
+            key in return.
+        html_parse_func (Optional[Callable]):
+            if "html_parse_func" is not None, then the
+            `html_parse_func` will be invoked with the response
+            text as input. The result is stored with
+            `self_define_func` key
```
Comment (DavdGao) [inline src/agentscope/service/web_search/web_digest.py:166]: There is no `self_define_func` key in lines 172-177.

**Review by DavdGao — COMMENTED** [review]
Please see inline comments, and solve the conflicts. 

Ps: 
We need to determine if we want developers to use the `parse_html` function directly. If so: 
1. The arguments `keep_raw` and `html_parse_func` in `parse_html` function are meaningless for developers when they use `parse_html` function directly. 
2. The "return raw" and "parse html by customized function" operations should be handled within the `load_web` function.

File: src/agentscope/service/web_search/web_digest.py (hunk 18 above)
Comment (ZiTao-Li) [reply src/agentscope/service/web_search/web_digest.py]: fixed

File: src/agentscope/service/web_search/web_digest.py (hunk 16 above)
Comment (ZiTao-Li) [reply src/agentscope/service/web_search/web_digest.py]: fixed

File: src/agentscope/service/web_search/web_digest.py (hunk 17 above)
Comment (ZiTao-Li) [reply src/agentscope/service/web_search/web_digest.py]: fixed

**Review by DavdGao — APPROVED** [review]
lgtm

**Pull request merged**